File: go/recursive/generate_parentheses.go
package main

//生成括号

//暴力法，找出所有的括号组合，并且一一判断是否有效
func generateParenthesis1(n int) []string {
	var res []string
	generateHelper1(0, 2*n, "", &res)
	return res
}

func generateHelper1(i int, n int, s string, res *[]string) {
	if i == n {
		if isValid(s) {
			*res = append(*res, s)
		}
		return
	}

	generateHelper1(i+1, n, s+")", res)
	generateHelper1(i+1, n, s+"(", res)
}

func isValid(s string) bool {
	flag := 0
	for _, v := range s {
		if v == '(' {
			flag++
		}
		if v == ')' {
			flag--
		}
		if flag < 0 {
			return false
		}
	}
	return flag == 0
}

//递归
func generateParenthesis2(n int) []string {
	var res []string
	generateHelper2(0, 0, n, "", &res)
	return res
}

func generateHelper2(left, right, n int, s string, res *[]string) {
	if len(s) == 2*n {
		*res = append(*res, s)
		return
	}

	if left < n {
		generateHelper2(left+1, right, n, s+"(", res)
	}
	if right < left {
		generateHelper2(left, right+1, n, s+")", res)
	}
}

File: java/algorithm/recursion/GenerateParentheses.java
package algorithm.recursion;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * @author roseduan
 * @time 2020/9/21 9:34 下午
 * @description 生成括号
 */
public class GenerateParentheses {

    /**
     * 第一种解法：暴力法，枚举所有的值，并且一一判断是否有效
     */
    public List<String> generateParenthesis1(int n) {
        List<String> res = new ArrayList<>();
        iterative(0, 2 * n, "", res);
        return res;
    }

    private void iterative(int i, int n, String res, List<String> list) {
        if (i == n) {
            if (isValid(res)) {
                list.add(res);
            }
            return;
        }

        iterative(i + 1, n, res + "(", list);
        iterative(i + 1, n, res + ")", list);
    }

    private boolean isValid(String str) {
        int i = 0;
        for (char c : str.toCharArray()) {
            if (c == '(') {
                i++;
            }
            if (c == ')') {
                i--;
            }
            if (i < 0) {
                return false;
            }
        }
        return i == 0;
    }

    /**
     * 第二种解法：使用回溯+剪枝的思想，在递归时，去掉不必要的条件循环
     */
    public List<String> generateParenthesis2(int n) {
        List<String> res = new ArrayList<>();
        helper(0,0, n,"", res);
        return res;
    }

    private void helper(int left, int right, int n, String res, List<String> list) {
        if (left == n && right == n) {
            list.add(res);
            return;
        }

        if (left < n) {
            helper(left + 1, right, n,res + "(", list);
        }
        if (right < left) {
            helper(left, right + 1, n,res + ")", list);
        }
    }

    /**
     * 第三种解法：迭代
     */
    public List<String> generateParenthesis3(int n) {
        List<List<String>> res = new ArrayList<>();
        res.add(Collections.singletonList(""));

        for (int i = 1; i <= n; i++) {
            List<String> list = new ArrayList<>();
            for (int j = 0; j < i; j++) {
                for (String first : res.get(j)) {
                    for (String second : res.get(i - 1 - j)){
                        list.add("(" + first + ")" + second);
                    }
                }
            }
            res.add(list);
        }
        return res.get(res.size() - 1);
    }
}
